py3status/modules/google_calendar.py

```diff
@@ -0,0 +1,489 @@
+
+Format_error placeholders:
+    {error} Error to display.
+
+Format_event & format_event_full placeholders:
+    {description} The description for the calendar event.
+    {end_date} The end date for the event.
+    {end_time} The end time for the event.
+    {location} The location for the event.
+    {start_date} The start date for the event.
+    {start_time} The start time for the event.
+    {summary} The summary (i.e. title) for the event.
+    {time_to} The time until the event starts (or until it is over
+        if already in progress).
+
+Format_time_to and format_time_left placeholders:
+    {days} The number of days until the event.
+    {hours} The number of hours until the event.
+    {minutes} The number of minutes until the event.
+
+Color options:
+    color_separator: Color for separator.
+    color_event: Color for a single event.
+    color_time: Color for the time associated with each event.
+
+Requires:
+    1. Python library google-api-python-client (can be obtained with pip).
+    2. OAuth 2.0 credentials for the Google Calendar api.
+
+    Follow Step 1 of the guide here to obtain your OAuth 2.0 credentials:
+    https://developers.google.com/google-apps/calendar/quickstart/python
+
+    Download the client_secret.json file which contains your client ID and
+    client secret. In your i3status config file, set configuration parameter
+    client_secret to the path to your client_secret.json file.
+
+    The first time you run the module, a browser window will open asking you
+    to authorize access to your calendar. After authorization is complete,
+    an access/refresh token will be saved to the path configured in
+    auth_token, and i3status will be restarted. This restart will
+    occur only once after the first time you successfully authorize.
+
+Examples:
+```
+# add color gradients for events and dates/times
+google_calendar {
+    thresholds = {
+        'event': [(1, '#d0e6ff'), (2, '#bbdaff'), (3, '#99c7ff'), (4, '#86bcff'),
+            (5, '#62a9ff'), (6, '#8c8cff'), (7, '#7979ff')],
+        'time': [(1, '#ffcece'), (2, '#ffbfbf'), (3, '#ff9f9f'), (4, '#ff7f7f'),
+            (5, '#ff5f5f'), (6, '#ff3f3f'), (7, '#ff1f1f')]
+    }
+}
+```
+
+@author Igor Grebenkov
+@license BSD
+"""
+
+import httplib2
+import os
+import datetime
+import pytz
+
+from apiclient import discovery
+from oauth2client import client
+from oauth2client import clientsecrets
+from oauth2client import tools
+from oauth2client.file import Storage
+from httplib2 import ServerNotFoundError
+
+
+SCOPES = 'https://www.googleapis.com/auth/calendar.readonly'
+APPLICATION_NAME = 'py3status google_calendar module'
+
+
+class Py3status:
+    auth_token = '~/.config/py3status/google_calendar.auth_token'
+    button_open = 3
+    button_refresh = 2
+    button_toggle = 1
+    cache_timeout = 60
+    client_secret = '~/.config/py3status/google_calendar.client_secret'
+    force_lower_case = False
+    format = '{events}|'
```

Comment:
We don't need `|`. Also, this should not be `{events}`, but rather something like `{format_event}`... which is a bit problematic here because you want something unbiassed like... `{events}` to juggle between `{format_event}` and `{format_event_full}`... Ideally, it'd be nice to merge them all.

Reply:
Oh right, 'cause we only need that to display some alternate text if `{events}` is `None`, right? I like how you used that little icon in `gcal` for that, so I'll make that the default in that case.

Reply:
Yes. Imperfect list... We get `else` on `None`, `''`, `[]`, etc... Sometimes we use `\?if=events Hi|Hey` for `0` and `False`. We can do lot of things with the basic formatting tricks. :-)

Reply:
Why `format_event` instead of `events`? It seems like the general convention for modules is `format_*` as a prefix for configuration parameters, but `{events}` is a placeholder?

Reply:
The `format_` prefix (in the config) denotes that we can use the formatting tricks here... and `format_` prefix (in the placeholder) denotes that it'll be something formatted.

We often use `format_` prefix (in the placeholder) when there can be more than one element... and it is usually done in the loop like the events here... because sometimes we can have more placeholders to choose from... and/or sometimes the placeholders can give us different values too... Doing this way will give users more freedom to customize... as we can let the users decide what to hide... or what to show... using basic formatting tricks.

Let me show you the future. Look at my recent `keyboard_locks` pull request where I'm getting rid of the icons in favor of putting everything in the format...

Now, let me show you the past. Look at current `dpms`. We have `icon_off`, `icon_on` there. Those config params used to be known as `format_off`, `format_on`... which is a bit confusing because we have `format` everywhere now... due to this pull request https://github.com/ultrabug/py3status/issues/98. Lot of things to catch up. 👍 

Future config...
```
dpms {
    # full customizability, no hardcoded.
    # can change all colors, texts, and format here.
    format = '\?if=is_enabled ON|OFF'
}
```
Current config...  Pull request to future: `404 Not Found` :-(
```
dpms {
    # full customizability, no hardcoded, lot of silly configs.
    # lot of configs to change all colors, texts, and format.
    format = `{icon}`
    icon_off = 'OFF'
    icon_on = 'ON'
}
```
Past config... Pull request to the current: https://github.com/ultrabug/py3status/pull/640
```
dpms {
    # partial customizability for text and colors, but not format.
    # format does not exist at this point. hardcoded.
    # fomat can't deal with prefix or suffix, only configs.
    format_on = `ON`
    format_off = `OFF`
}
```
This does not exist, but if it does... It wouldn't make sense.
```
dpms {
    #  same as above, but can cause confusion. we use icon.
    format = '\?if=dpms {format_on}|{format_off}'
    format_on = `ON`
    format_off = `OFF`
}
```
This may be somewhat to what you're doing here.
```
dpms {
    # partial customizability, where is '{on}' coming from? hardcoded
    format = '\?if=dpms {on}'
    format_on = `ON`
    format_off = `OFF`
}
```
I hope that helps a lot. :-)

Reply:
That does help a lot, thanks! That actually gives me an idea for how to fold `format_event` and `format_event_full` into one, since I can use something `\?if=format_full ...` to show extended info.

Reply:
`\?if=is_toggled` would work here too.

Reply:
I don't think that would allow for toggling more than one event at a time, though, since it would be applied to all events.